src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,500 @@
+        {
+            var (document, textSpan, cancellationToken) = context;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var action = await IntroduceParameterAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (action != null)
+            {
+                context.RegisterRefactoring(action, textSpan);
+            }
+        }
+
+        public async Task<CodeAction?> IntroduceParameterAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var expression = await document.TryGetRelevantNodeAsync<TExpressionSyntax>(textSpan, cancellationToken).ConfigureAwait(false);
+            if (expression == null || CodeRefactoringHelpers.IsNodeUnderselected(expression, textSpan))
+            {
+                return null;
+            }
+
+            var semanticDocument = await SemanticDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+            var expressionType = semanticDocument.SemanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (expressionType is null or IErrorTypeSymbol)
+            {
+                return null;
+            }
+
+            if (expression != null)
+            {
+                var (title, actions) = AddActions(semanticDocument, expression, cancellationToken);
+
+                if (actions.Length > 0)
+                {
+                    return new CodeActionWithNestedActions(title, actions, isInlinable: true);
+                }
+            }
+
+            return null;
+        }
+
+        public async Task<Solution> IntroduceParameterForRefactoringAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+
+            var annotatedExpression = new SyntaxAnnotation(ExpressionAnnotationKind);
+
+            var annotatedSemanticDocument = await GetAnnotatedSemanticDocumentAsync(document, annotatedExpression, expression, cancellationToken).ConfigureAwait(false);
+            var annotatedExpressionWithinDocument = (TExpressionSyntax)annotatedSemanticDocument.Root.GetAnnotatedNodesAndTokens(annotatedExpression).Single().AsNode()!;
+            var methodSymbolInfo = GetMethodSymbolFromExpression(annotatedSemanticDocument, annotatedExpressionWithinDocument, cancellationToken);
+            var methodCallSites = await FindCallSitesAsync(annotatedSemanticDocument, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+
+            var updatedCallSitesSolution = await RewriteCallSitesAsync(annotatedSemanticDocument, annotatedExpressionWithinDocument, methodCallSites, cancellationToken).ConfigureAwait(false);
+
+            if (updatedCallSitesSolution == null)
```

Comment:
can this happen?  does RewriteCallSitesAsync ever return null?

Reply:
Yes, if there is the case you want to introduce a parameter and no method callsites actually exist, then I assume the expected behavior would still be to refactor the method with the new parameter?

Reply:
sure.  hwever, i would not expect the updated solution to be null. it woudl just be the same isntance as the original solution.

Reply:
Oh true, I can just return the old solution